doc/compiler/tril/TrilLanguageReference.md

```diff
@@ -0,0 +1,228 @@
+has been created in Tril to make writing test cases easier. 
+
+# Basics
+
+Tril is built out of two main components: 
+
+1. A very permissive S-expression parser.
+2. An IlGenerator that walks the AST produced by said parser, and generates
+   Trees. 
+
+# Syntax
+
+Tril represents Trees using S-expressions, with key-value properties to augment
+a particular node with more information. 
+
+* `(x (y) (z))` represents some `x` with `y` and `z` as children.
+* `(x foo=bar baz=braz (y) (z)` is the same tree as above, but where the node
+  for `x` is augmented with two properties with values. 
+
+  Anything may be provided as a property for a node, however, unknown
+  properties will be silently ignored.
+
+## Literals
+
+* Integer literals can be input directly, and will always be treated internally
+  as a 64-bit integer.  
+* Hexadecimal floating point literals are supported as well, and are treated as
+  64-bit integers. Use a leading `0x`. 
+* Floating point numbers can be input by including a decimal point.  
+
+## Comments 
+
+`;` starts a comment in tril, and the comment proceeds to the end of a line. 
+
+## Special Forms: 
+
+The key to Tril is the unlocking of extra behaviour using 'special forms', or
+atoms that represent certain behaviours in Testarossa. 
+
+### Methods
+
+A whole method in Tril is wrapped in a `method` expression. 
+
+#### Properties 
+
+* `return` _Mandatory_ is the return type of the method. 
+* `args` _Optional_ is a square-bracketed list of arguments representing the
+  types of the parameters, and the order in which they appear. 
+
+  eg. `args=[Int32,Double,Float]`
+    
+### Block
+
+In the Testarossa IL, all trees exist within the context of a basic block. In 
+Tril, blocks must be children of a `method`, or they will not be recognized.
+
+#### Properties
+
+* `name` _Optional_ Blocks can be named in order to target them with branches. 
+
+### Stores and loads
+
+Currently the creation of new symbol references in Tril is not supported.
+However, store and load opcodes provide two options for where the store / load
+can come from, which can be specified with one of the following two proprties: 
+
+* `parm` can specify the parameter from which a load can come. Parameters are 
+  counted from zero.  eg. `(iload parm=2)` loads the third parameter from the
+  signature. 
+* `temp` Stores to temps create new temporary variables that can later be
+  referenced. eg. `(istore temp="x" ...)`
+
+### Branches 
+
+Branches evaluate a condition then either jump to a target, or falls through. 
+
+#### Properties
+
+* `target` _Mandatory_ The `name` of the block a successful computation of the
+  condition will continue executing in. 
+* `fallthrough` _Optional_ In some circumstances it is necessary to explicitly
+  list the `name` of the fallthrough block of a branch to ensure the desired CFG is
+  constructed. In addition to a block `name`, there are two special values that
```

Comment:
"...two optional special values..."   (add optional)

Reply:
Not sure if the "optional" makes sense. In my mind, these are values that carry a special meaning; they assign names to targets that are otherwise unnamed. They are not elements that can be "optionally" added. Maybe an example using these would help clarify this: `fallthrough="foo"` vs `fallthrough=@exit`.

Reply:
Yeah, on second thought "optional" is not the right word here and makes more sense without it.  +1 to an example though.

Reply:
I actually don't yet have enough experience to write out an example for `fallthrough=@exit`; 

    (goto target="unconditionalTarget" fallthrough=@none) 

makes sense to me -- not sure though, is that a mandatory fallthrough on an unconditional branch or not?

Reply:
A throw with a fallthrough of @exit would make sense with a target being some catch block, but not sure if that is implemented yet or if the @exit in that case would be implicit?

Reply:
I don't think we've got any support for `throw` -- I mean, if it worked due to the cleverness of the way that @Leonardo2718 cooked up Tril, I wouldn't be surprised, but we definitely haven't tried it 🤔

Reply:
So, it used to be that specifying `fallthrough` was mandatory in Tril. This created the need to have a way of expressing "no fall-through" (`@none`) and "fall-through to the exit block" (`@exit`).

However, it was later decided that in most cases, it is easy enough to figure out what the fall-through path should be (the next block in TreeTop order). So, the `fallthrough` property was made optional. Now, Tril will implicitly generate fall-through edges as needed by walking the AST, *unless an explicit `fallthrough` is specified*. Explicitly specifying the expected (correct) fall-through target (as is done in @mgaudet's example above) is therefore redundant.

Today, `fallthrough` is used to *override* the default fall-through. So far, this has been most useful for explicitly generating incorrect CFGs for testing purposes.

re: throw, I don't know enough about how exceptions work in Testarossa to say whether or not they will work. Quickly scanning the code base, it doesn't look like Tril will correctly handle throws. At least, I don't think it will generate the correct CFG (though it might generate the correct Trees; cleverness can only go so far 😛 ).

In any case, the documentation and examples should make the context under which `@none` and `@exit` are/should be used clear.